### PR TITLE
Implement irradiance volumes.

### PR DESCRIPTION
**Pull request by pcwalton**
# Objective

Bevy could benefit from *irradiance volumes*, also known as *voxel global illumination* or simply as light probes (though this term is not preferred, as multiple techniques can be called light probes).
Irradiance volumes are a form of baked global illumination; they work by sampling the light at the centers of each voxel within a cuboid. At runtime, the voxels surrounding the fragment center are sampled and interpolated to produce indirect diffuse illumination.

## Solution

This is divided into two sections. The first is copied and pasted from the irradiance volume module documentation and describes the technique. The second part consists of notes on the implementation.

### Overview

An *irradiance volume* is a cuboid voxel region consisting of
regularly-spaced precomputed samples of diffuse indirect light. They're
ideal if you have a dynamic object such as a character that can move about
static non-moving geometry such as a level in a game, and you want that
dynamic object to be affected by the light bouncing off that static
geometry.

To use irradiance volumes, you need to precompute, or *bake*, the indirect
light in your scene. Bevy doesn't currently come with a way to do this.
Fortunately, [Blender] provides a [baking tool] as part of the Eevee
renderer, and its irradiance volumes are compatible with those used by Bevy.
The [`bevy-baked-gi`] project provides a tool, `export-blender-gi`, that can
extract the baked irradiance volumes from the Blender `.blend` file and
package them up into a `.ktx2` texture for use by the engine. See the
documentation in the `bevy-baked-gi` project for more details as to this
workflow.

Like all light probes in Bevy, irradiance volumes are 1×1×1 cubes that can
be arbitrarily scaled, rotated, and positioned in a scene with the
[`bevy_transform::components::Transform`] component. The 3D voxel grid will
be stretched to fill the interior of the cube, and the illumination from the
irradiance volume will apply to all fragments within that bounding region.

Bevy's irradiance volumes are based on Valve's [*ambient cubes*] as used in
*Half-Life 2* ([Mitchell 2006], slide 27). These encode a single color of
light from the six 3D cardinal directions and blend the sides together
according to the surface normal.

The primary reason for choosing ambient cubes is to match Blender, so that
its Eevee renderer can be used for baking. However, they also have some
advantages over the common second-order spherical harmonics approach:
ambient cubes don't suffer from ringing artifacts, they are smaller (6
colors for ambient cubes as opposed to 9 for spherical harmonics), and
evaluation is faster. A smaller basis allows for a denser grid of voxels
with the same storage requirements.

If you wish to use a tool other than `export-blender-gi` to produce the
irradiance volumes, you'll need to pack the irradiance volumes in the
following format. The irradiance volume of resolution *(Rx, Ry, Rz)* is
expected to be a 3D texture of dimensions *(Rx, 2Ry, 3Rz)*. The unnormalized
texture coordinate *(s, t, p)* of the voxel at coordinate *(x, y, z)* with
side *S* ∈ *{-X, +X, -Y, +Y, -Z, +Z}* is as follows:

```text
s = x

t = y + ⎰  0 if S ∈ {-X, -Y, -Z}
        ⎱ Ry if S ∈ {+X, +Y, +Z}

        ⎧   0 if S ∈ {-X, +X}
p = z + ⎨  Rz if S ∈ {-Y, +Y}
        ⎩ 2Rz if S ∈ {-Z, +Z}
```

Visually, in a left-handed coordinate system with Y up, viewed from the
right, the 3D texture looks like a stacked series of voxel grids, one for
each cube side, in this order:

| **+X** | **+Y** | **+Z** |
| ------ | ------ | ------ |
| **-X** | **-Y** | **-Z** |

A terminology note: Other engines may refer to irradiance volumes as *voxel
global illumination*, *VXGI*, or simply as *light probes*. Sometimes *light
probe* refers to what Bevy calls a reflection probe. In Bevy, *light probe*
is a generic term that encompasses all cuboid bounding regions that capture
indirect illumination, whether based on voxels or not.

Note that, if binding arrays aren't supported (e.g. on WebGPU or WebGL 2),
then only the closest irradiance volume to the view will be taken into
account during rendering.

[*ambient cubes*]: https://advances.realtimerendering.com/s2006/Mitchell-ShadingInValvesSourceEngine.pdf

[Mitchell 2006]: https://advances.realtimerendering.com/s2006/Mitchell-ShadingInValvesSourceEngine.pdf

[Blender]: http://blender.org/

[baking tool]: https://docs.blender.org/manual/en/latest/render/eevee/render_settings/indirect_lighting.html

[`bevy-baked-gi`]: https://github.com/pcwalton/bevy-baked-gi

### Implementation notes

This patch generalizes light probes so as to reuse as much code as possible between irradiance volumes and the existing reflection probes. This approach was chosen because both techniques share numerous similarities:

1. Both irradiance volumes and reflection probes are cuboid bounding regions.
2. Both are responsible for providing baked indirect light.
3. Both techniques involve presenting a variable number of textures to the shader from which indirect light is sampled. (In the current implementation, this uses binding arrays.)
4. Both irradiance volumes and reflection probes require gathering and sorting probes by distance on CPU.
5. Both techniques require the GPU to search through a list of bounding regions.
6. Both will eventually want to have falloff so that we can smoothly blend as objects enter and exit the probes' influence ranges. (This is not implemented yet to keep this patch relatively small and reviewable.)

To do this, we generalize most of the methods in the reflection probes patch #11366 to be generic over a trait, `LightProbeComponent`. This trait is implemented by both `EnvironmentMapLight` (for reflection probes) and `IrradianceVolume` (for irradiance volumes). Using a trait will allow us to add more types of light probes in the future. In particular, I highly suspect we will want real-time reflection planes for mirrors in the future, which can be easily slotted into this framework.

## Changelog

> This section is optional. If this was a trivial fix, or has no externally-visible impact, you can delete this section.

### Added
* A new `IrradianceVolume` asset type is available for baked voxelized light probes. You can bake the global illumination using Blender or another tool of your choice and use it in Bevy to apply indirect illumination to dynamic objects.

### CODE REVIEW

**Comment by pcwalton** [comment]
This needs a bit of cleaning up and documentation, but it works and is pretty close.

**Comment by JMS55** [comment]
Leaving this in the milestone for now, but we might have to end up cutting it depending on time constraints. @pcwalton if this is ready to go, make sure to take it out of draft.

**Comment by pcwalton** [comment]
This should be ready to go now. I've verified that this works on Windows, macOS, Android, and WebGL 2. The `check-bans` CI failure seems unrelated.

File: crates/bevy_pbr/src/light_probe/irradiance_volume.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+//! workflow.
+//!
+//! Like all light probes in Bevy, irradiance volumes are 1×1×1 cubes that can
+//! be arbitrarily scaled, rotated, and positioned in a scene with the
+//! [`bevy_transform::components::Transform`] component. The 3D voxel grid will
+//! be stretched to fill the interior of the cube, and the illumination from the
+//! irradiance volume will apply to all fragments within that bounding region.
+//!
+//! Bevy's irradiance volumes are based on Valve's [*ambient cubes*] as used in
+//! *Half-Life 2* ([Mitchell 2006], slide 27). These encode a single color of
```
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/irradiance_volume.rs:27]: You can link directly to this by appending `#page=27` to the url btw

File: crates/bevy_pbr/src/light_probe/irradiance_volume.rs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+//! geometry.
+//!
+//! To use irradiance volumes, you need to precompute, or *bake*, the indirect
+//! light in your scene. Bevy doesn't currently come with a way to do this.
+//! Fortunately, [Blender] provides a [baking tool] as part of the Eevee
+//! renderer, and its irradiance volumes are compatible with those used by Bevy.
+//! The [`bevy-baked-gi`] project provides a tool, `export-blender-gi`, that can
+//! extract the baked irradiance volumes from the Blender `.blend` file and
+//! package them up into a `.ktx2` texture for use by the engine. See the
+//! documentation in the `bevy-baked-gi` project for more details as to this
```
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/irradiance_volume.rs:17]:
```suggestion
//! documentation in the `bevy-baked-gi` project for more details on this
```

File: examples/3d/irradiance_volumes.rs [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+const FOX_SCALE: f32 = 0.05;
+const SPHERE_SCALE: f32 = 2.0;
+
+const IRRADIANCE_VOLUME_INTENSITY: f32 = 150.0;
+
+static DISABLE_IRRADIANCE_VOLUME_HELP_TEXT: &str = "Press Space to disable the irradiance volume";
+static ENABLE_IRRADIANCE_VOLUME_HELP_TEXT: &str = "Press Space to enable the irradiance volume";
+
+static STOP_ROTATION_HELP_TEXT: &str = "Press Enter to stop rotation";
```
Comment (JMS55) [inline examples/3d/irradiance_volumes.rs:31]: Can you format the text to be like `(Space): [Enable/Disable] he irradiance volume`

File: crates/bevy_pbr/src/render/mesh_view_bindings.wgsl [hunk 4]
```diff
@@ -46,37 +46,45 @@
 #endif
 @group(0) @binding(15) var environment_map_sampler: sampler;
 
-@group(0) @binding(16) var dt_lut_texture: texture_3d<f32>;
-@group(0) @binding(17) var dt_lut_sampler: sampler;
+#ifdef MULTIPLE_LIGHT_PROBES_IN_ARRAY
+@group(0) @binding(16) var irradiance_volumes: binding_array<texture_3d<f32>, 8u>;
```
Comment (JMS55) [inline crates/bevy_pbr/src/render/mesh_view_bindings.wgsl:50]: I'm worried we're going to go over the texture bindings on minspec devices. For transmission it was already causing issues. With irradiance volumes, reflection probes, and lightmaps, we're probably blowing past the limits...

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl [hunk 5]
```diff
@@ -1,16 +1,21 @@
...
 
+#import bevy_pbr::light_probe::query_light_probe
 #import bevy_pbr::mesh_view_bindings as bindings
 #import bevy_pbr::mesh_view_bindings::light_probes
 
 struct EnvironmentMapLight {
     diffuse: vec3<f32>,
     specular: vec3<f32>,
+    // Whether any applicable reflection probe was found.
+    found: bool,
```
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/environment_map.wgsl:11]: I don't think we need the found field. You can just initialize the diffuse/specular to 0 when nothing is found, and return early in other cases. Cleaner and a bit less branching.

File: crates/bevy_pbr/src/light_probe/irradiance_volume.rs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+//! | ------ | ------ | ------ |
+//! | **-X** | **-Y** | **-Z** |
+//!
+//! A terminology note: Other engines may refer to irradiance volumes as *voxel
+//! global illumination*, *VXGI*, or simply as *light probes*. Sometimes *light
+//! probe* refers to what Bevy calls a reflection probe. In Bevy, *light probe*
+//! is a generic term that encompasses all cuboid bounding regions that capture
+//! indirect illumination, whether based on voxels or not.
+//!
+//! Note that, if binding arrays aren't supported (e.g. on WebGPU or WebGL 2),
```
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/irradiance_volume.rs:71]: Please link to the wgpu feature here.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Can you add a toggle to some kind of visualization of the irradiance volume (gizmos or meshes) to the example/

File: crates/bevy_pbr/src/render/mesh_view_bindings.wgsl (hunk 4 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_view_bindings.wgsl:50]: I was concerned about this too, but it did work on the Android simulator and WebGL 2. Is there some documentation on how to test on a min-spec device?

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl (hunk 5 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:11]: Good point, I didn't realize that you might as well ignore a light probe that provides zero light. I'll do that.

File: crates/bevy_pbr/src/render/mesh_view_bindings.wgsl (hunk 4 above)
Comment (JMS55) [reply crates/bevy_pbr/src/render/mesh_view_bindings.wgsl:50]: I think you can configure the wgpu Limits as part of the WgpuSettings when configuring the plugins. Otherwise you can also just modify the initialize_render() function from bevy_render to use the minimum/default wgpu limits.

Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_view_bindings.wgsl:50]: What is the minimum limit that Bevy supports?

**Comment by pcwalton** [comment]
@JMS55 So I added the toggle requested, but it doesn't show up that well in the example because the irradiance volume ends above the camera. I tried drawing a small circle at the center of every voxel but that was waaaaaay too slow in debug mode.

**Comment by pcwalton** [comment]
I reduced the number of voxels in the irradiance volume to a more reasonable level, so now the gizmo shows the center of each voxel.

**Comment by pcwalton** [comment]
@JMS55 Actually, you know what, I only count 12 textures that are enabled in the PBR shader, even with this PR. So I think we're under the 16 limit.

**Comment by pcwalton** [comment]
I added a visualization mode to the example that shows the voxels.

**Comment by pcwalton** [comment]
This should be ready for re-review. All comments have been addressed.

File: assets/shaders/irradiance_volume_voxel_visualization.wgsl [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+@fragment
+fn fragment(mesh: VertexOutput) -> @location(0) vec4<f32> {
+    // Using `irradiance_volumes[0]` is a little dodgy, but we know we only have
+    // one irradiance volume in this scene, so it's fine.
+    //
+    // `irradiance_volume_light()` multiplies by intensity, so cancel it out.
+    // If we take intensity into account, these spheres will be way too bright.
+    let rgb = irradiance_volume::irradiance_volume_light(
+        mesh.world_position.xyz,
```
Comment (rodolphito) [inline assets/shaders/irradiance_volume_voxel_visualization.wgsl:13]: this will not sample the exact center of the voxel will it? it might be worthwhile to do some coordinate snapping math to ensure the exact color stored in the voxel is reproduced

File: crates/bevy_pbr/src/light_probe/light_probe.wgsl [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        if (all(abs(probe_space_pos) <= vec3(0.5f))) {
+            result.texture_index = light_probe.cubemap_index;
+            result.intensity = light_probe.intensity;
+            result.inverse_transform = inverse_transform;
+
+            // TODO: Workaround for ICE in DXC https://github.com/microsoft/DirectXShaderCompiler/issues/6183
+            // This works because it's the last thing that happens in the for loop, and will break as soon as it
+            // goes back to the top of the loop.
+            // We can't use `break` here because of the ICE.
+            // break;
```
Comment (rodolphito) [inline crates/bevy_pbr/src/light_probe/light_probe.wgsl:62]: this is missing the line that early outs
```suggestion
            // break;
            light_probe_index = light_probe_count;
```

File: crates/bevy_pbr/src/light_probe/mod.rs [hunk 9]
```diff
@@ -119,24 +172,123 @@ pub struct LightProbesBuffer(DynamicUniformBuffer<LightProbesUniform>);
...
+{
+    /// The list of environment maps presented to the shader, in order.
+    binding_index_to_textures: Vec<C::AssetId>,
+
+    /// The reverse of `binding_index_to_cubemap`: a map from the texture ID to
+    /// the index in `binding_index_to_cubemap`.
+    cubemap_to_binding_index: HashMap<C::AssetId, u32>,
+
+    /// Information about each light probe, ready for upload to the GPU, sorted
+    /// in order from closest to the camera to farthest.
```
Comment (rodolphito) [inline crates/bevy_pbr/src/light_probe/mod.rs:228]: Shouldnt this be sorted by screen coverage? If a light probe is really close to the camera but behind it, its not useful. If a light probe is really large and covers a lot of the screen but its center is far from the camera, it should still be more/as important as small lightprobes that are closer by.

This can be handled in a followup.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl [hunk 10]
```diff
@@ -377,11 +410,10 @@ fn apply_pbr_lighting(
...
 #endif
 
-#ifdef LIGHTMAP
-    indirect_light += in.lightmap_light * diffuse_color;
-#endif
+    // Ambient light (indirect)
+    indirect_light += ambient::ambient_light(in.world_position, in.N, in.V, NdotV, diffuse_color, F0, perceptual_roughness, diffuse_occlusion);
 
-    let emissive_light = emissive.rgb * output_color.a;
+    var emissive_light = emissive.rgb * output_color.a;
```
Comment (rodolphito) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:416]: why was this changed?

File: examples/3d/irradiance_volumes.rs [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+        };
+
+        let resolution = image.texture_descriptor.size;
+        let scale = vec3(
+            1.0 / resolution.width as f32,
+            1.0 / resolution.height as f32,
+            1.0 / resolution.depth_or_array_layers as f32,
+        );
+
+        // Spawn a sphere for each voxel.
```
Comment (rodolphito) [inline examples/3d/irradiance_volumes.rs:563]:
```suggestion
        // Spawn a cube for each voxel.
```

File: examples/3d/irradiance_volumes.rs [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+
+fn create_spheres(
+    image_assets: Res<Assets<Image>>,
+    mut commands: Commands,
+    irradiance_volumes: Query<(&IrradianceVolume, &GlobalTransform)>,
+    voxel_cube_parents: Query<Entity, With<VoxelCubeParent>>,
+    voxel_cubes: Query<Entity, With<VoxelCube>>,
+    example_assets: Res<ExampleAssets>,
+) {
+    // If voxel spheres have already been spawned, don't do anything.
```
Comment (rodolphito) [inline examples/3d/irradiance_volumes.rs:542]:
```suggestion
    // If voxel cubes have already been spawned, don't do anything.
```

File: examples/3d/irradiance_volumes.rs [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+
+// Plays the animation on the fox.
+fn play_animations(assets: Res<ExampleAssets>, mut players: Query<&mut AnimationPlayer>) {
+    for mut player in players.iter_mut() {
+        // This will safely do nothing if the animation is already playing.
+        player.play(assets.fox_animation.clone()).repeat();
+    }
+}
+
+fn create_spheres(
```
Comment (rodolphito) [inline examples/3d/irradiance_volumes.rs:534]:
```suggestion
fn create_cubes(
```

File: examples/3d/irradiance_volumes.rs [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+        }))
+        .add_plugins(MaterialPlugin::<VoxelVisualizationMaterial>::default())
+        .init_resource::<AppStatus>()
+        .init_resource::<ExampleAssets>()
+        .insert_resource(AmbientLight {
+            color: Color::WHITE,
+            brightness: 0.0,
+        })
+        .add_systems(Startup, setup)
+        .add_systems(PreUpdate, create_spheres)
```
Comment (rodolphito) [inline examples/3d/irradiance_volumes.rs:136]:
```suggestion
        .add_systems(PreUpdate, create_cubes)
```

**Review by rodolphito — COMMENTED** [review]
Tested and the light fringing i had observed before is fixed. Awesome work!

File: crates/bevy_pbr/src/light_probe/mod.rs (hunk 9 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/mod.rs:228]: Yes, eventually what we want to do is to use clustering just like we use for lights. This logic is a temporary hack for now to reduce the complexity of this patch.

File: crates/bevy_pbr/src/light_probe/light_probe.wgsl (hunk 8 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/light_probe.wgsl:62]: This should be handled by the `result.texture_index < 0` in the condition.

Comment (rodolphito) [reply crates/bevy_pbr/src/light_probe/light_probe.wgsl:62]: please update the comment to explain this

File: crates/bevy_pbr/src/render/pbr_functions.wgsl (hunk 10 above)
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl]: Removed it.

File: assets/shaders/irradiance_volume_voxel_visualization.wgsl (hunk 7 above)
Comment (pcwalton) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:29]: OK, I do that now.

**Comment by pcwalton** [comment]
Comments addressed and ready for re-review.

**Review by rodolphito — APPROVED** [review]
Looks great!

File: assets/shaders/irradiance_volume_voxel_visualization.wgsl [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    // middle of the nearest texel.
+    var unit_pos = (irradiance_volume_info.inverse_transform *
+        vec4(mesh.world_position.xyz, 1.0f)).xyz;
+    let resolution = vec3<f32>(irradiance_volume_info.resolution);
+    let stp = clamp((unit_pos + 0.5) * resolution, vec3(0.5f), resolution - vec3(0.5f));
+    let stp_rounded = round(stp - 0.5f) + 0.5f;
+    let rounded_world_pos = (irradiance_volume_info.transform * vec4(stp_rounded, 1.0f)).xyz;
+
+    // `irradiance_volume_light()` multiplies by intensity, so cancel it out.
+    // If we take intensity into account, these spheres will be way too bright.
```
Comment (JMS55) [inline assets/shaders/irradiance_volume_voxel_visualization.wgsl:27]:
```suggestion
    // If we take intensity into account, the cubes will be way too bright.
```

File: assets/shaders/irradiance_volume_voxel_visualization.wgsl [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    inverse_transform: mat4x4<f32>,
+    resolution: vec3<u32>,
+    intensity: f32,
+}
+
+@group(2) @binding(100)
+var<uniform> irradiance_volume_info: VoxelVisualizationIrradianceVolumeInfo;
+
+@fragment
+fn fragment(mesh: VertexOutput) -> @location(0) vec4<f32> {
```
Comment (JMS55) [inline assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: Why do the cubes appear to be gray+shaded? I was thinking we would want to show essentially an "unlit" solid-color cube the color of the probe.

Comment (pcwalton) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: They are doing what you suggested, I'm fairly sure. I don't see any shading.

I believe the gray comes from the grayish background color, which is roughly the same in Blender so the indirect light from it is captured in the probe. The sides of the cubes will be a mix of the gray from the background with the colors of the floor visible from that cube side. The relative mix of gray vs. floor color is based on how much that side of the cube can "see".

Comment (JMS55) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: So the irradiance volumes are sampling the clear color/background? Seems strange. ClearColor shouldn't be used for lighting. Only a Skybox if specified, but not the clear color itself.

Comment (pcwalton) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: In reality it's sampling the "world" in Blender which happens to be dark gray by default. Would you prefer that I change it to black?

Comment (JMS55) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: Yeah I think that would be better. Or drop in a skybox and use that.

Comment (pcwalton) [reply assets/shaders/irradiance_volume_voxel_visualization.wgsl:16]: OK, I added a skybox and re-baked the irradiance volume. It looks right to my eyes.

**Comment by pcwalton** [comment]
Screenshot 2024-02-02 145444

**Comment by pcwalton** [comment]
Don't merge yet. The CI failure is legit.

**Review by rodolphito — APPROVED** [review]

**Comment by pcwalton** [comment]
This should be ready for review again.

File: assets/shaders/irradiance_volume_voxel_visualization.wgsl [hunk 17]
```diff
@@ -0,0 +1,33 @@
+#import bevy_pbr::forward_io::VertexOutput
+#import bevy_pbr::irradiance_volume
+#import bevy_pbr::mesh_view_bindings
+
+struct VoxelVisualizationIrradianceVolumeInfo {
+    transform: mat4x4<f32>,
+    inverse_transform: mat4x4<f32>,
+    resolution: vec3<u32>,
+    intensity: f32,
```
Comment (NthTensor) [inline assets/shaders/irradiance_volume_voxel_visualization.wgsl:9]: This is in Lux, yes? Good to note the unit here.

File: crates/bevy_pbr/src/light_probe/environment_map.rs [hunk 18]
```diff
@@ -184,6 +163,19 @@ pub(crate) enum RenderViewBindGroupEntries<'a> {
...
+/// a global environment map that lights everything visible in the view, as
+/// opposed to a light probe which affects only a specific area.
+pub struct EnvironmentMapViewLightProbeInfo {
+    /// The index of the diffuse and specular cubemaps in the binding arrays.
+    pub(crate) cubemap_index: i32,
+    /// The smallest mip level of the specular cubemap.
+    pub(crate) smallest_specular_mip_level: u32,
+    /// The scale factor applied to the diffuse and specular light in the
+    /// cubemap.
+    pub(crate) intensity: f32,
```
Comment (NthTensor) [inline crates/bevy_pbr/src/light_probe/environment_map.rs:176]: Units again here please. Is this Lux or Watt per square meter?

**Review by NthTensor — APPROVED** [review]
